src/ekat/ekat_physical_constants.hpp

```diff
@@ -0,0 +1,105 @@
+
+    // Constructor for single value (i.e., NIST std)
+    PhysicalConstant(const RealType& nvalue, const ekat::units::Units& units, const std::string& name,
+      const std::string& info) :
+      m_nist_val(nvalue), m_e3sm_val(nvalue), m_units(units), m_name(name), m_info(info) {}
+
+
+    // Constructor for two different values (i.e., when E3SM value and NIST value are different)
+    PhysicalConstant(const RealType& evalue, const RealType& nvalue, const ekat::units::Units& units,
+      const std::string& name) : m_e3sm_val(evalue), m_nist_val(nvalue), m_units(units),
+      m_name(name), m_info("") {}
+
+    // Constructor for two different values (i.e., when E3SM value and NIST value are different)
+    PhysicalConstant(const RealType& evalue, const RealType& nvalue, const ekat::units::Units& units,
+      const std::string& name, const std::string& info) : m_e3sm_val(evalue), m_nist_val(nvalue), m_units(units),
+      m_name(name), m_info(info) {}
+
+
+    // Constructor for a constant that is defined by the product or quotient of 2 other constants.
+    PhysicalConstant(const PhysicalConstant<RealType>& lhs, const Op& op, const PhysicalConstant<RealType>& rhs,
+       const std::string& name) :  m_e3sm_val(lhs.m_e3sm_val), m_nist_val(lhs.m_nist_val),
+       m_units(lhs.m_units), m_name(name), m_info(lhs.m_info + "; " + rhs.m_info) {
+        switch (op) {
+          case (Times) : {
+            m_e3sm_val *= rhs.m_e3sm_val;
+            m_nist_val *= rhs.m_nist_val;
+            m_units = lhs.m_units * rhs.m_units;
+            break;
+          }
+          case (Divide) : {
+            m_e3sm_val /= rhs.m_e3sm_val;
+            m_nist_val /= rhs.m_nist_val;
+            m_units = lhs.m_units / rhs.m_units;
+            break;
+          }
+        }
+    }
+
+    std::string name() const {return m_name;}
+
+    std::string get_string() const {return m_name + " " + to_string(m_units) +
+      (m_info.empty() ? "" : " (" + m_info + ")");}
+
+    std::string unit_string() const {return to_string(m_units);}
+
+    RealType value() const {return m_e3sm_val;}
+
+  // allow subclasses for submodules (e.g., shoc, p3, haero, etc.)
+  protected:
+    RealType m_nist_val;
+    RealType m_e3sm_val;
+    ekat::units::Units m_units;
+
+    std::string m_name;
+    std::string m_info;
+
+    friend bool operator == (const PhysicalConstant&, const PhysicalConstant&);
+};
+
+template <typename RealType>
+inline bool operator == (const PhysicalConstant<RealType>& lhs, const PhysicalConstant<RealType>& rhs) {
+  return lhs.m_e3sm_val == rhs.m_e3sm_val &&
+         lhs.m_nist_val == rhs.m_nist_val &&
+         lhs.m_units == rhs.m_units;
+}
+
+template <typename RealType>
+inline bool operator != (const PhysicalConstant<RealType>& lhs, const PhysicalConstant<RealType>& rhs) {
+  return !(lhs == rhs);
+}
+
+// ================== COMMON PHYSICAL CONSTANTS =============== //
```

Comment:
I would put all these constants in a separate namespace, something like `physical_constants`. The way it is now, they pollute the ekat namespace.

Also, I would put them in a separate file, with "double" in the filename, so people know these constants have a pre-defined precision.

Reply:
Agree completely.  I put them here just for the sake of this PR, to see if people are ok with them being namespace variables (in their own namespace) or if people want them wrapped up in a struct.

Reply:
I see +/- of either approach:

- namespace: you can "add" to a namespace in a separate file, but then the namespace (and file) need to make clear what precision is used.
- struct: whether sp or dp is clear from the struct template arg, but you cannot extend the struct.

Which one to pick, is a matter of taste, I think. Personally, I'd probably go with a struct, and have apps define their own namespace. But I don't feel really strong about it.

One thing I'd do, though, is to explicitly type `<double>` in the lines below, just in case someone changes the class default (impossible for downstream app to detect) or remove the default altogether (more likely, although would immediately generate compiler errors).

Reply:
Added a separate file and namespace for common constants; it's what `ekat_units.hpp` does, and it's closer to the fortran domain scientists are familiar with already.  No strong feelings though --- happy to encapsulate in a templated struct if others prefer it.   On that note...

Changed all constants to double precision.  Is there harm in this choice? (params can still compute in single)

Reply:
> Changed all constants to double precision. Is there harm in this choice? (params can still compute in single)

Not sure what you mean. You mean having 14 digits when hard-coding stuff like this?
```
const auto Pi = PhysicalConstant<>(3.14159265358979323846264, ekat::units::Units::nondimensional(), "Pi");
```
That's fine.

Reply:
No I meant, is there harm in `using RealType = double;` (with regard to single-precision builds)

Reply:
I don't think so, at least not necessarily. But the user need to be aware of that, or else hard-to-track non bfb issues might emerge. Take this:

```
float my_mass = 1.0/3.0; // or some other non-exact sp floating point
auto force = PhysicalConstants<>::g_accel * my_mass;
```
The variable `force` will be a double, due to C++ promotion rules. And even forcing the result to be a float might not help:
```
float my_mass = 1.0/3.0; // or some other non-exact sp floating point
float my_length = 1.0/3.0; // or some other non-exact sp floating point
float work1 = (PhysicalConstants<>::g_accel * my_mass) * my_length;
float work2 = (PhysicalConstants<float>::g_accel * my_mass) * my_length;
```
The intermediate result in parentheses for work has double precision, so my_length will be temporarily promoted to double for the second product. The result will likely not be bfb with work2.

Again, that's fine, as long as users are making responsible use of the default. E.g., in scream, if they wrap the constant inside a `sp(...)` call, then all is good.

Reply:
Actually, in my 1st example, even with `float` instead of `auto` you might get quite a different output. Without carries, you might expect 0-1 ulp error. With carries, anything is on the table (say your constant is 1+epsilon<float>/2, and you add epsilon<float>/2. If the constant is first converted to float, you get 1, otherwise you get 1+epsilon<float>.